### PR TITLE
Product list preview in product list V2

### LINKED ISSUE #30212: Product page v2 list not showing product if its not in category "Home"
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add screenshots

Product page v2 list not showing product if its not in category "Home".
It seems to be due to the QueryBuilder which defaults to filter by category id of PS_HOME_CATEGORY in PrestaShop\PrestaShop\Core\Grid\Query\ProductQueryBuilder::getFilteredCategoryId

### Expected behavior

As in product page v1 it should show all the products that exists in database.

### Steps to reproduce

1. Make sure experimental product page v2 is on (Advanced Parameters -> New &Experimental features -> New product page - Single store switched on).
2. Go to BO products list and open any product for edition and scroll down to categories part
3. Remove "Home" category (you may first need to change the default category, because you cannot remove current default product category). This step could probably also be done in product page v1 by unselecting home category.
4. Go to products list and see that the product is gone from list. (you could still find it by filtering by other category though)

### PrestaShop version(s) where the bug happened

develop

### PHP version(s) where the bug happened

_No response_

### If your bug is related to a module, specify its name and its version

_No response_

### PR DESCRIPTION
**Pull request by jolelievre**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | 8.0.x
| Description?      | This PR adds the preview action in the product list v2, we had to introduce an intermediate Symfony action because the service that generates the URL isn't usable directly in the grid (which only works based on Symfony routes)
| Type?             | new feature
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes #30212
| Related PRs       | ~
| How to test?      | This PR is the last one for the product list V2 which has not been fully tested yet, so it should be used as an opportunity to test it as a whole
| Possible impacts? | ~

### CODE REVIEW

**Review by matks — DISMISSED** [review]

**Review by marsaldev — DISMISSED** [review]

**Review by khouloudbelguith — COMMENTED** [review]
Hi @jolelievre,

**Issue1**
I followed this spec: https://build.prestashop.com/prestashop-specs/1.7/back-office/catalog/products/listing-v2.html#multistore

If the product row has more than one shop in the Store(s) column, the preview button `should not be displayed` in `All shops / Groups shops`
image
image

 **Issue2**

We can set the `product_id_product_min_field` > `product_id_product_max_field`

https://user-images.githubusercontent.com/16067358/180038599-603cf763-7b32-42f5-90d7-72414ed221f7.mp4

Product V1 => OK, I attached a screen record

https://user-images.githubusercontent.com/16067358/180038784-733e0cea-be1b-45e4-8ddf-2eb9ce2433e5.mp4

**Issue3**
When we resize the window to 1400px * 980px, The IDs input are not well displayed
image

https://user-images.githubusercontent.com/16067358/180040152-e21d3a12-61d6-4b3a-b838-383eab7e0e78.mp4


**Issue4**
We cannot filter with a negative quantity 
image

**Issue5**

1. Go to BO > Product list V2
2. Click to filter by category & select a category, for example, `Clothes`
3. Search by position => OK
4. Click on Unselect "Filter by category"
5. The search by position is not reset, we have only one product displayed & no filter displayed

image
I attached a screen record

https://user-images.githubusercontent.com/16067358/180042974-88891f11-e7d5-4d60-b1a0-13d57d6cc197.mp4

PS, we should click on Unselect twice to reset the filter.

**Issue6**
The button `Clear filter` is missing comparing product V1
image
image

**Issue7**

1. Go to BO > Product V2
2. Create a disabled product & empty name
3. Go to BO > Product V2 list & try to enable
4. See error 500 in the console & no alert displayed


https://user-images.githubusercontent.com/16067358/180045540-8e992b43-a52e-4352-9c9a-a33c9e82888b.mp4

**Issue8**

1. Create 21 products
2. In the paginations select 20 items per page
3. Go to the second page & delete the product displayed with **Bulk action**
4. See error => You are not redirected to the first page

https://user-images.githubusercontent.com/16067358/180047346-4091b19c-bd24-4d21-80f2-330d5bbfc3b8.mp4

Thanks!

**Comment by jolelievre** [comment]
Hi @khouloudbelguith and @MatShir 

I'm starting to work on these bugs, only the first one is about multishop and should not be part of this PR's scope (or even the version's scope). So I opened a dedicated issue https://github.com/PrestaShop/PrestaShop/issues/29709

And I added one more issue to this PR that @Hlavtox discovered 

### Issue 9
1. Create a product that is not associated with the root category
2. Display the product page with no Category filters selected
3. The product is not visible

There seems to be a default filter by root category when none is selected, there shouldn't be this filter when no category is selected.

**Comment by jolelievre** [comment]
Hi @khouloudbelguith 

I handled your feedback:

**Issue2**
Technically it's not related to this page only, it's a global issue related to Grid because we never performed any validation on the filters. So I still added a quick management of this problem, filters are validated and if invalid not applied and an error message is displayed after refresh. Since it's a new feature I won't go deeper into this, if it needs to be improved it should be done in a future dedicated issue I only handle the minimal use case.

@MatShir I'd like to show you the result just so you can validate if it's acceptable (see the video):

https://user-images.githubusercontent.com/13801017/193307794-ac7ba2c4-31cd-44b3-a47d-ddc5d24626b7.mp4

**Issue3**
I don't think it's related to this Grid specifically we have similar issues in many other grid and it's related to general UI of the BO, when we use table we don't really handle the cells size which are adapted based on the content which can vary a lot depending on the headers and the data itself. So this should be hanled on a higher level to decide how a grid should be displayed when it's too large.

**Issue 4**
Fixed.

**Issue 5**
Fixed. About your PS note:
> PS, we should click on Unselect twice to reset the filter.
Like the following issue, this has not been clearly defined from a UX point of view so it should be done in a future issue. For now it's functional enough IMHO (considering this is still an experimental page which is gonna be improved in 8.1)

**Issue6**
The button is present but not at the same place indeed. Since we didn't have a clear design mockup for this feature I don't think we should block the PR for this. But instead, create an issue to improve this in 8.1.0.

https://user-images.githubusercontent.com/13801017/193315461-6ff58f2d-7843-4f0a-97ef-8e2f8c9352c5.mp4

**Issue 7**
Fixed, and the toggle is switched back to its initial state automatically.

**Issue 8**
It is not a bug related to the specific product page itself but a bug related to the internal behaviour of ALL grids. This has already been handled in this PR https://github.com/PrestaShop/PrestaShop/pull/28782 so I won't fix it here.

**Issue 9**
Fixed.

**Comment by Hlavtox** [comment]
Issue 3 is not related to Jo's PR, we need to fix it separately. We can improve the situation by reducing input paddings, the we will see what we can do. 👍 I will discuss with @NeOMakinG.

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php [hunk 1]
```diff
@@ -140,6 +141,27 @@ public function lightListAction(ProductFilters $filters, Request $request): Resp
         ]);
     }
 
+    /**
+     * The redirection URL is generation thanks to the ProductPreviewProvider however it can't be used in the grid
```
Comment (zuk3975) [inline src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php:248]:
```suggestion
     * The redirection URL is generated by the ProductPreviewProvider, but it cannot be used in the grid
```

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php [hunk 2]
```diff
@@ -140,6 +141,27 @@ public function lightListAction(ProductFilters $filters, Request $request): Resp
...
+     * since the LinkRowAction expects a symfony route, so this action is merely used as a proxy for symfony routing
```
Comment (zuk3975) [inline src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php:249]: have you considered adding an option in the LinkRowAction to toggle between route and direct url?

File: src/PrestaShopBundle/Form/Admin/Type/NumberMinMaxFilterType.php [hunk 3]
```diff
@@ -47,6 +49,22 @@ public function configureOptions(OptionsResolver $resolver)
...
             'min_field_options' => [],
             'max_field_options' => [],
+            'constraints' => [
+                new Callback([
+                    'callback' => function (?array $impactData, ExecutionContextInterface $context) {
+                        if (!isset($impactData['min_field']) || !isset($impactData['max_field'])) {
+                            return;
+                        }
+
+                        if ($impactData['min_field'] > $impactData['max_field']) {
```
Comment (zuk3975) [inline src/PrestaShopBundle/Form/Admin/Type/NumberMinMaxFilterType.php:59]: isn't there a risk for inaccurate float comparisons? Maybe we should use DecimalNumber here to compare? (also maybe then we could reuse the same constraint in both integer and number MinMaxFilterTypes?

**Review by zuk3975 — DISMISSED** [review]
couple suggestions to consider

**Review by matks — DISMISSED** [review]
Nothing to add to @zuk3975 suggestions

File: src/PrestaShopBundle/Form/Admin/Type/NumberMinMaxFilterType.php (hunk 3 above)
Comment (jolelievre) [reply src/PrestaShopBundle/Form/Admin/Type/NumberMinMaxFilterType.php:59]: Yeah I guess it would be wprth a refacto to integrate those improvements, if there are some bugs found I'll do it in this PR but I prefer not to book it for now

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php (hunk 2 above)
Comment (jolelievre) [reply src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php:249]: Yeah but the issue is that the rows actions/urls need to be generated differently on each row so that it uses the row's ID For now it's easily done in twig thanks to the `path` twig function and just by adapting the route parameters

But here it would be more complicated as we would need to be able to use the `ProductPreviewProvider` service from twig, by creating a dedicated filter/function I guess, and then it should be configurable to use either router or alternative generation function, it's not impossible but this solution seemed more simple 😅

**Review by khouloudbelguith — COMMENTED** [review]
Hello @jolelievre,

**Issue1**

1. Go to BO > Catalog > Product V2
2. Click on "Filter by categories" & select any category (for example Art)
3. Sort by Position
4. Click on reset
5. See exception

image

https://user-images.githubusercontent.com/16067358/197823193-a36fa225-bfe2-48a1-8ba8-ab4cca56f605.mp4

**Without PR** => OK


**Issue2**
1. Go to BO > Catalog > Product V2
2. Filter by Status "Yes"
3. Try to disable any product with quick edit (for example the first product)
4. The filter result is not automatically updated
5. Click on reset
6. See the status of the product is just enabled

I attached a screen record

https://user-images.githubusercontent.com/16067358/197824317-162ff1ac-c825-4dfd-ab6a-79c288493422.mp4

I checked the Product page V1 => OK

https://user-images.githubusercontent.com/16067358/197827035-aa11a8a1-a7ba-43c8-9d34-b2e82a47bc2a.mp4

**Issue3**
About this [issue](https://github.com/PrestaShop/PrestaShop/issues/29709), I still reproduced it with this PR 
image

I attached a screen 

https://user-images.githubusercontent.com/16067358/197832631-e2ff980f-ad87-4e7c-b26d-e0aef3e9f271.mp4


**Without PR** I checked with develop & 8.0.x => OK  
image


Thanks!

**Comment by jolelievre** [comment]
Hi @khouloudbelguith 

thanks for your feedback.

**Issue 1**
Fixed.

**Issue 2**
This is not a bug, this grid simply uses a different component for this grid. The toggle column is handled by ajax (based on the `AsyncToggleColumnExtension` js component, the purpose is to perform the toggle action via a JS call, so refreshing the list after makes no sense it would lose the whole purpose of this component. Besides it's not the only grid: categories, customers, states, ... behave the same.

But we also have some other grids (like Language grid) that use `ColumnTogglingExtension` in this case the action is performed like a form and the whole page is updated (and so are the filtered value of course).

TLDR; these are two valid ways to handle toggle actions so it's not a bug, it's just two different features.

**Issue 3**
I know the issue is still there, but it WON'T be fixed by this PR because it's about multishop, and multishop list will only be developed in 8.1 version. In the other branches, there is no bug because the preview action is not here at all, it was added in this current PR only for now. That's why [I reopened the issue](https://github.com/PrestaShop/PrestaShop/issues/29709) that was closed by mistake.

**Review by zuk3975 — DISMISSED** [review]

**Review by NeOMakinG — DISMISSED** [review]
front lgtm

**Review by khouloudbelguith — COMMENTED** [review]
Hello @jolelievre,

**Issue1**
I found a difference between the product pages V1 & V2

1. Go to BO > Catalog > Product V2
2. Click on "Filter by categories" & select any category (for example Art)
3. Try to filter by any filter for example quantity & click on the search
4. Click on reset
5. the filter by category will be reset & the filter by quantity is reset too

I attached a screen record

https://user-images.githubusercontent.com/16067358/200623932-9097f027-d855-4266-bb7d-00d502b9fc53.mp4


In the product V1, only the filter by quantity will be reset

https://user-images.githubusercontent.com/16067358/200623952-e73fd417-2c35-4eab-8a33-d4922f3fdd8d.mp4

Ping @MatShir what do you think?

I tried to follow this spec: https://build.prestashop-project.org/prestashop-specs/1.7/back-office/catalog/products/listing-v2.html, but I found a lot of differences

**Issue2**
https://build.prestashop-project.org/prestashop-specs/1.7/back-office/catalog/products/listing-v2.html#empty-state
_Spec_: `The empty state card appears WHEN there is no product`

Install this PR

1. Go to BO > Advanced Parameters > New & Experimental Features & enable New product page - Single store
2. Delete all products
3. See error => No card displayed 
image
4. Click on Filter by category > select any category
5. See error > all filters are displayed, they shouldn't


https://user-images.githubusercontent.com/16067358/200625481-abc2a5b8-0455-4afa-81ff-640568b2bff1.mp4

**Issue3**
https://build.prestashop-project.org/prestashop-specs/1.7/back-office/catalog/products/listing-v2.html#top-of-the-listing
_Spec_: `No Confirmation Modal before applying the action "Activate action, Deactivate action, and Duplicate action".`

 In my case, when using those actions, I have a confirmation modal

for example, Activate action, => 
image


**Issue4**
https://build.prestashop-project.org/prestashop-specs/1.7/back-office/catalog/products/listing-v2.html#top-of-the-listing
_Spec_: `An alert success is displayed after applying those actions "Activate selection, Deactivate selection, Duplicate action and Delete action".`

In my case, I don't have any alert
Activate selection => I have
image
Deactivate action => I have
image
Duplicate action => I have
image
Delete action => I have
image


**Issue5**
When we filter by a category, we don't have the Rearrange button and the "Save & refresh".
This part is  not the same that I have
_Spec_

```php
A column position is added in the table between status and actions with a “Rearrange” button allowing to change the position of the products in the selected category.
Rearrange button:
After clicking on it:

If a sort and/or search were done they are removed and products are sorted by position in the selected category.
The rearrange button becomes a “Save & refresh” disabled button
It’s possible to change the position of the products in the category by clicking on the line or the icon
Changing the position of products:
After clicking on the icon or the line, the checkbox is checked and the product is selected (issue [15291](https://build.prestashop-project.org/prestashop-specs/1.7/back-office/catalog/products/[https://github.com/PrestaShop/PrestaShop/issues/15291](https://github.com/PrestaShop/PrestaShop/issues/15291)))

When I change the position of one or many products:

Products checkboxes are disabled
Its new position is indicated in red in the listing
The “Save & refresh” button is enabled
After clicking on “Save & refresh”:

New positions are saved and a confirmation message is displayed “Products successfully sorted”
If search were done before changing positions, they are kept
```

**Issue6**
https://build.prestashop-project.org/prestashop-specs/1.7/back-office/catalog/products/listing-v2.html#columns
_Spec_: `The alt of the image is the caption saved on the product page, if the caption is empty the alt is the name of the product in the current language (https://github.com/PrestaShop/PrestaShop/issues/10236)
`
In our case, No alt displayed
image


In the product page V1, we have an alt but it is empty
image


**Issue7**
https://build.prestashop-project.org/prestashop-specs/1.7/back-office/catalog/products/listing-v2.html#links
_Spec_: `Image: Open the product in the same browser tab and on the basic settings tab`

In my case, when I click on the image => No redirection to the Basic setting of the product


_Spec_: `Quantity: If the product has combinations, open it in the same browser tab on the combinations tab`

In my case, redirection to the Basic Settings of the product but not the combinations Tab

[Issue7.webm](https://user-images.githubusercontent.com/16067358/200632226-0f2ca696-d783-4fb8-b82f-a6aa7d6180cc.webm)

Thanks!

**Comment by MatShir** [comment]
> Hello @jolelievre,
> 
> **Issue1** I found a difference between the product pages V1 & V2
> 
> 1. Go to BO > Catalog > Product V2
> 2. Click on "Filter by categories" & select any category (for example Art)
> 3. Try to filter by any filter for example quantity & click on the search
> 4. Click on reset
> 5. the filter by category will be reset & the filter by quantity is reset too
> 
> I attached a screen record
> 
>  _Product.V2.mp4 
> In the product V1, only the filter by quantity will be reset
> 
>  Products.PrestaShop.v1.mp4 
> Ping @MatShir what do you think?
> 
> I tried to follow this spec: https://build.prestashop-project.org/prestashop-specs/1.7/back-office/catalog/products/listing-v2.html, but I found a lot of differences

To be fixed → It should only reset the filter from the grid and keep the filter from the category

> **Issue2** https://build.prestashop-project.org/prestashop-specs/1.7/back-office/catalog/products/listing-v2.html#empty-state _Spec_: `The empty state card appears WHEN there is no product`
> 
> Install this PR
> 
> 1. Go to BO > Advanced Parameters > New & Experimental Features & enable New product page - Single store
> 2. Delete all products
> 3. See error => No card displayed
>    image
> 4. Click on Filter by category > select any category
> 5. See error > all filters are displayed, they shouldn't
> 
>  Issue2.mp4 

Display the page "add your first product" -> like in the V1.

> **Issue3** https://build.prestashop-project.org/prestashop-specs/1.7/back-office/catalog/products/listing-v2.html#top-of-the-listing _Spec_: `No Confirmation Modal before applying the action "Activate action, Deactivate action, and Duplicate action".`
> 
> In my case, when using those actions, I have a confirmation modal
> 
> for example, Activate action, => image

Right behavior → Need to update the spec

> **Issue4** https://build.prestashop-project.org/prestashop-specs/1.7/back-office/catalog/products/listing-v2.html#top-of-the-listing _Spec_: `An alert success is displayed after applying those actions "Activate selection, Deactivate selection, Duplicate action and Delete action".`
> 
> In my case, I don't have any alert Activate selection => I have image Deactivate action => I have image Duplicate action => I have image Delete action => I have image
> 

Need to update the specs. Thanks !

> **Issue5** When we filter by a category, we don't have the Rearrange button and the "Save & refresh". This part is not the same that I have _Spec_
> 
> ```
> A column position is added in the table between status and actions with a “Rearrange” button allowing to change the position of the products in the selected category.
> Rearrange button:
> After clicking on it:
> 
> If a sort and/or search were done they are removed and products are sorted by position in the selected category.
> The rearrange button becomes a “Save & refresh” disabled button
> It’s possible to change the position of the products in the category by clicking on the line or the icon
> Changing the position of products:
> After clicking on the icon or the line, the checkbox is checked and the product is selected (issue [15291](https://build.prestashop-project.org/prestashop-specs/1.7/back-office/catalog/products/[https://github.com/PrestaShop/PrestaShop/issues/15291](https://github.com/PrestaShop/PrestaShop/issues/15291)))
> 
> When I change the position of one or many products:
> 
> Products checkboxes are disabled
> Its new position is indicated in red in the listing
> The “Save & refresh” button is enabled
> After clicking on “Save & refresh”:
> 
> New positions are saved and a confirmation message is displayed “Products successfully sorted”
> If search were done before changing positions, they are kept
> ```
> 

To respect the position feature on the product. I need to update (save and refresh is not used anymore)

> **Issue6** https://build.prestashop-project.org/prestashop-specs/1.7/back-office/catalog/products/listing-v2.html#columns _Spec_: `The alt of the image is the caption saved on the product page, if the caption is empty the alt is the name of the product in the current language (https://github.com/PrestaShop/PrestaShop/issues/10236) ` In our case, No alt displayed image
> 
> In the product page V1, we have an alt but it is empty image

If the product doesn't have a name, then the alt is N/A.
Inject in the tag alt the product title in the employee language.
If there is no image, display no image path / "no image available" in the list (in the language of the employee).

> **Issue7** https://build.prestashop-project.org/prestashop-specs/1.7/back-office/catalog/products/listing-v2.html#links _Spec_: `Image: Open the product in the same browser tab and on the basic settings tab`
> 
> In my case, when I click on the image => No redirection to the Basic setting of the product
> 
> _Spec_: `Quantity: If the product has combinations, open it in the same browser tab on the combinations tab`
> 
> In my case, redirection to the Basic Settings of the product but not the combinations Tab
> 
>  Issue7.webm 
> Thanks!

The quantity link in a row for product combination redirects to the 'Combination Tab'
The ref link redirects to the 'Detail Tab'

**Review by NeOMakinG — DISMISSED** [review]

**Review by zuk3975 — DISMISSED** [review]

**Review by zuk3975 — APPROVED** [review]

**Review by NeOMakinG — APPROVED** [review]

**Review by khouloudbelguith — APPROVED** [review]
Hi @jolelievre,

```
I found only one issue
https://build.prestashop-project.org/prestashop-specs/1.7/back-office/catalog/products/listing-v2.html#columns Spec: The alt of the image is the caption saved on the product page, if the caption is empty the alt is the name of the product in the current language (https://github.com/PrestaShop/PrestaShop/issues/10236)In our case, the alt is displayed but do not check that the caption is empty yes or no
``` 

=> Issue fixed

It is ok :heavy_check_mark: 

No product is created => OK
image

I tried when "Enable stock management" is disabled, the quantity column is not displayed
image

1. I tried when Taxes were disabled
2. Sort by all fields ASC & DESC
3. Filter by all fields ( + sort + pagination)
4. Check Bulk actions
5. Filter by categories
6. Try to preview disabled / enabled product
7. Try to delete / duplicate a product
8. Try to import products with csv =>well displayed
9. Try to add a product with API => well displayed

Thanks!

**Comment by jolelievre** [comment]
Thanks @khouloudbelguith
